File: Sort/QuickSort.py
class Solution:
    """
    @param A: an integer array
    @return: nothing
    """

    def Partition(self, L, low, high):
        pivot_key = L[low]
        while (low < high):
            while low < high and L[high] >= pivot_key:
                high -= 1
            L[low] = L[high]

            while low < high and L[low] <= pivot_key:
                low += 1
            L[high] = L[low]

        L[low] = pivot_key
        return low

    def QSort(self, L, low, high):
        m = low + (high - low) // 2
        if L[low] > L[high]:
            L[low], L[high] = L[high], L[low]
        if L[m] > L[high]:
            L[m], L[high] = L[high], L[m]
        if L[low] < L[m]:
            L[low], L[m] = L[m], L[low]

        if (high - low) > 7:
            while low < high:
                pivot = self.Partition(L, low, high)
                self.QSort(L, low, pivot)
                low = pivot + 1

            return
        
        self.InsertSort(L)

    def InsertSort(self, L):
        length = len(L)
        i = 1
        while i < length:
            if L[i-1] > L[i]:
                tmp = L[i]
                j = i - 1
                while j >=0 and L[j] > tmp:
                    L[j+1] = L[j]
                    j -= 1
                L[j+1] = tmp

            i += 1
        return

    def sortIntegers(self, A):
        # write your code here
        if A == []:
            return []
        self.QSort(A, 0, len(A)-1)

s = Solution()
A = [3, 2, 1, 4, 5]
s.sortIntegers(A)
print(A)

File: Sort/QuickSort.c
#include <stdio.h>
#include "InsertSort.c"
#define MAX_LENGTH_INSERT_SORT 7

int Partition(SqList *L, int low, int high)
{
    int pivotkey = L->r[low];
    while (low < high)
    {
        // find number smaller than pivot
        while (low < high && L->r[high] >= pivotkey)
            high--;
        L->r[low] = L->r[high];

        // find number bigger than pivot
        while (low < high && L->r[low] <= pivotkey)
            low++;
        L->r[high] = L->r[low];
    }
    L->r[low] = pivotkey; // L->r[high] = pivotkey;
    return low; // return high;
}

void QSort(SqList *L, int low, int high)
{
    int m = low + (high - low) / 2;
    if (L->r[low] > L->r[high])
    {
        swap(L, low, high);
    }
    if (L->r[m] > L->r[high])
    {
        swap(L, m, high);
    }
    if (L->r[low] < L->r[m])
    {
        swap(L, m, low);
    }

    int pivot;
    if ((high-low) > MAX_LENGTH_INSERT_SORT)
    {
        while (low < high)
        {
            pivot = Partition(L, low, high);
            QSort(L, low, pivot);
            low = pivot + 1;
        }

        return;
    }

    InsertSort(L);
}

void QuickSort(SqList *L)
{
    QSort(L, 0, L->length - 1);
}

// int main()
// {
//     SqList L = {
//         {50, 10, 90, 30, 70, 40, 80, 60, 20},
//         9,
//     };

//     QuickSort(&L);
//     for (int i = 0; i < L.length; i++)
//         printf("%d ", L.r[i]);
// }